### ISSUE TITLE
Issue with TrailingComma Cop

### ISSUE THREAD

**Issue by hackling** [bbatsov/rubocop#2369]
Here is the example spec I wrote for `spec/rubocop/cop/style/trailing_comma_spec.rb`

```ruby
    it 'accepts a method with an object with a multiline method call' do
      inspect_source(cop, <<-CODE)
        some_method(
          SomeObject
            .method
            .method
        )
      CODE
      expect(cop.offenses).to be_empty
    end
```

Test Output
```
Failures:

  1) RuboCop::Cop::Style::TrailingComma with single line list of values when EnforcedStyleForMultiline is comma accepts a method with an object with a multiline method call
     Failure/Error: expect(cop.offenses).to be_empty
       expected `[#<RuboCop::Cop::Offense:0x007f88ebb5a8f8 @severity=#<RuboCop::Cop::Severity:0x007f88ebb5a8d0 @name=:convention>, @location=#<Parser::Source::Range (string) 74...81>, @message="Put a comma after the last parameter of a multiline method call.", @cop_name="Style/TrailingComma", @corrected=false>].empty?` to return true, got false
     Shared Example Group: "single line lists" called from ./spec/rubocop/cop/style/trailing_comma_spec.rb:121
     # ./spec/rubocop/cop/style/trailing_comma_spec.rb:48:in `block (3 levels) in <top (required)>'

Finished in 13.66 seconds (files took 1.52 seconds to load)
3443 examples, 1 failure, 3 pending

Failed examples:

rspec ./spec/rubocop/cop/style/trailing_comma_spec.rb[1:1:2:5] # RuboCop::Cop::Style::TrailingComma with single line list of values when EnforcedStyleForMultiline is comma accepts a method with an object with a multiline method call

Randomized with seed 9208
```

I would expect it to detect that while I am multilining, I only have one parameter, and the need for a comma here is not relevant.

Alternatively, it would be great if this `TrailingComma` cop could be configured to apply to collection objects, or methods, or both.

**Comment by alexdowad** [comment]
Thanks for pointing this out. Just opened a PR which fixes the problem.

**Comment by bbatsov** [comment]
> Alternatively, it would be great if this TrailingComma cop could be configured to apply to collection objects, or methods, or both.

Totally. I was always opposed to applying this on method calls.

**Comment by alexdowad** [comment]
> I was always opposed to applying this on method calls.

Are there people who like it on method calls??

**Comment by bbatsov** [comment]
If memory serves - yes. :-)

**Comment by alexdowad** [comment]
Do you want another PR which adds a config parameter to disable it for method calls? Let me guess: yes.

**Comment by bbatsov** [comment]
Indeed. And let's disable this check by default. :-)

**Comment by hackling** [comment]
This is awesome. I started poking around yesterday, but the turn around on this is a lot quicker than I was expecting. :+1:

**Comment by alexdowad** [comment]
I'm just trying to figure out what the extra config param should be called. Or should `TrailingComma` be broken up into `TrailingCommaCollection` and `TrailingCommaMethodCall`?

**Comment by bbatsov** [comment]
I'm fine either way.

On Friday, 30 October 2015, Alex Dowad <notifications@github.com> wrote:

> I'm just trying to figure out what the extra config param should be
> called. Or should TrailingComma be broken up into TrailingCommaCollection
> and TrailingCommaMethodCall?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bbatsov/rubocop/issues/2369#issuecomment-152649335>.
>


-- 
Best Regards,
Bozhidar Batsov

http://www.batsov.com

**Comment by alexdowad** [comment]
Oh no, you mean I have to make a decision??? :sweat_smile:

**Comment by bbatsov** [comment]
I don't make decisions in the absence of an actual keyboard. :-)

On Saturday, 31 October 2015, Alex Dowad <notifications@github.com> wrote:

> Oh no, you mean I have to make a decision??? [image: :sweat_smile:]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bbatsov/rubocop/issues/2369#issuecomment-152727409>.
>


-- 
Best Regards,
Bozhidar Batsov

http://www.batsov.com

**Comment by ClayShentrup** [comment]
This is wrong. For instance, I might change this:

```ruby
foo.fetch(
  :bar,
)
```

To this:

```ruby
foo.fetch(
  :bar,
  nil,
)
```

I don't want the diff to show that the second line changed.

Moreover, this isn't logically consistent. It shouldn't matter how many other arguments came before the last argument. So if you think this is correct:

```ruby
foo.fetch(
  :bar
)
```

Then you have to argue that this is also correct:

```ruby
foo.fetch(
  :bar,
  nil
)
```

**Comment by hackling** [comment]
Sure, which is why I think being able to turn this on/off for methods is a good idea.

@alexdowad If you need help deciding, I think splitting it up into two cops is better than a param.

**Comment by bbatsov** [comment]
Guess @hackling is right. Examples like the ones given by @ClayShentrup are relatively uncommon, but they are still valid and I guess we should retain as much flexibility as possible, so that everyone's happy at the end of the day.

**Comment by ryanswood** [comment]
I too think along the lines of @ClayShentrup and prefer to use commas in all cases. Right now, there does not seem to be a config option to disable the new multi-line method chain sole arg exception so upgrading to the newest Rubocop version is not an option without refactoring.

 Is there an active plan to separate the `TrailingComma` cop into 2 separate cops? If I am the only one who is wanting this configurability, I can *possibly* find some time in the distant future...

**Comment by alexdowad** [comment]
@ryanswood, I definitely have an active plan to do this. At the moment, I am working on fixing up Racc and `unparser` -- sorry that I haven't cleared this one yet.

**Comment by ryanzbone** [comment]
@ryanswood, you and @ClayShentrup are definitely not the only ones who wants this

**Comment by mkantor** [comment]
I find it particularly strange that these are treated differently:

```ruby
foo(
  bar: 'baz',
)
```

```ruby
foo(
  bar,
)
```

Especially now that Ruby has keyword arguments.

**Comment by alexdowad** [comment]
Everybody, sorry for taking a long time to address this issue. I have just pushed a commit to my open PR which reverses the unfortunate change made in response to this issue. Since we have separate `Style/TrailingCommaInArguments` and `Style/TrailingCommaInLiteral` cops now, @hackling can disable trailing commas in method calls, but keep them in hash and array literals, if that is what he actually wants.

Or if @hackling thinks that single-argument method calls are special enough that we should have a style which treats them differently, that should be a new style for `Style/TrailingCommaInArguments` and `Style/TrailingCommaInLiteral`, rather than breaking the existing `comma` and `consistent_comma` styles.
